caffe2/proto/torch.proto

```diff
@@ -4,547 +4,85 @@ import "caffe2/proto/caffe2.proto";
-// versions.
-//
-// Here are the most notable contortions we have to carry out to work around
-// these limitations:
-//
-//   - No 'map' (added protobuf 3.0). We instead represent mappings as lists
-//     of key-value pairs, where order does not matter and duplicates
-//     are not allowed.
-
-// Versioning
-//
-// ONNX versioning is specified in docs/IR.md and elaborated on in docs/Versioning.md
-//
-// To be compatible with both proto2 and proto3, we will use a version number
-// that is not defined by the default value but an explicit enum number.
-enum Version {
-  // proto3 requires the first enum value to be zero.
-  // We add this just to appease the compiler.
+enum ProtoVersion {
   _START_VERSION = 0;
-  // The version field is always serialized and we will use it to store the
-  // version that the  graph is generated from. This helps us set up version
-  // control.
-  // For the IR, we are using simple numbers starting with with 0x00000001,
-  // which was the version we published on Oct 10, 2017.
-  IR_VERSION_2017_10_10 = 0x0000000000000001;
-
-  // IR_VERSION 2 published on Oct 30, 2017
-  // - Added type discriminator to AttributeProto to support proto3 users
-  IR_VERSION_2017_10_30 = 0x0000000000000002;
-
-  // IR VERSION 3 published on Nov 3, 2017
-  // - For operator versioning:
-  //    - Added new message OperatorSetIdProto
-  //    - Added opset_import in ModelProto
-  // - For vendor extensions, added domain in NodeProto
-  IR_VERSION_NEWEST_ONNX = 0x0000000000000003;
-
-  // PYTORCH IR VERSION
-  IR_VERSION_NEWEST = 0x0000000000000103;
+  IR_VERSION_NEWEST = 0x0000000000000101;
 }
 
-// Attributes
-//
-// A named attribute containing either singular float, integer, string, graph,
-// and tensor values, or repeated float, integer, string, graph, and tensor values.
-// An AttributeProto MUST contain the name field, and *only one* of the
-// following content fields, effectively enforcing a C/C++ union equivalent.
-message AttributeProto {
-
-  // Note: this enum is structurally identical to the OpSchema::AttrType
-  // enum defined in schema.h.  If you rev one, you likely need to rev the other.
-  enum AttributeType {
-    UNDEFINED = 0;
-    FLOAT = 1;
-    INT = 2;
-    STRING = 3;
-    TENSOR = 4;
-    GRAPH = 5;
-
-    FLOATS = 6;
-    INTS = 7;
-    STRINGS = 8;
-    TENSORS = 9;
-    GRAPHS = 10;
+message MethodDef {
+  // method name
+  optional string name = 1; // method name
+
+  // static graph
+  optional caffe2.NetDef graph = 2;
+  // method is represented as torch script
+  optional string torch_script = 3;
+
+  // the names of inputs and outputs
+  repeated string inputs = 4;
+  repeated string outputs = 5;
+
+  // method type, can be one of the following:
+  //   1) MAIN, the main method, by default is `forward` method
+  //   2) INIT, only run once for initialization
+  //   3) REGULAR, may run multiple times at runtime
+  enum MethodType {
```

Comment:
I do not like this. First, there is no equivalent in PyTorch, so we will never generate anything but REGULAR method. Second, all the behavior can be completely described by the method name "forward" is the main method "init" is the initialization method.  Having this enum here will encourage a proliferation of hacky different method types. All methods should execute the same way.

Reply:
Agree, will use method name to indicate the type of method.